core/matrix/fbcsr.cpp

```diff
@@ -0,0 +1,513 @@
+
+        for (int ibr = 0; ibr < bs; ibr++)
+            for (int jbr = 0; jbr < bs; jbr++)
+                values(cur_bnz, ibr, jbr) = it->second(ibr, jbr);
+
+        if (it->first.block_row > cur_brow) {
+            tmp->row_ptrs_.get_data()[++cur_brow] = cur_bnz;
+        } else {
+            assert(cur_brow == it->first.block_row);
+            assert(cur_bcol <= it->first.block_column);
+        }
+
+        cur_bcol = it->first.block_column;
+        cur_bnz++;
+    }
+
+    tmp->row_ptrs_.get_data()[++cur_brow] =
+        static_cast<index_type>(blocks.size());
+
+    assert(cur_brow == tmp->get_size()[0] / bs);
+
+    tmp->move_to(this);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::write(mat_data &data) const
+{
+    std::unique_ptr<const LinOp> op{};
+    const Fbcsr *tmp{};
+    if (this->get_executor()->get_master() != this->get_executor()) {
+        op = this->clone(this->get_executor()->get_master());
+        tmp = static_cast<const Fbcsr *>(op.get());
+    } else {
+        tmp = this;
+    }
+
+    data = {tmp->get_size(), {}};
+
+    const blockutils::DenseBlocksView<const value_type, index_type> vblocks(
+        tmp->values_.get_const_data(), bs_, bs_);
+
+    for (size_type brow = 0; brow < tmp->get_num_block_rows(); ++brow) {
+        const auto start = tmp->row_ptrs_.get_const_data()[brow];
+        const auto end = tmp->row_ptrs_.get_const_data()[brow + 1];
+
+        for (auto inz = start; inz < end; ++inz) {
+            for (int ib = 0; ib < bs_; ib++) {
+                const auto row = brow * bs_ + ib;
+                for (int jb = 0; jb < bs_; jb++) {
+                    const auto col =
+                        tmp->col_idxs_.get_const_data()[inz] * bs_ + jb;
+                    const auto val = vblocks(inz, ib, jb);
+                    data.nonzeros.emplace_back(row, col, val);
+                }
+            }
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Fbcsr<ValueType, IndexType>::transpose() const
+{
+    auto exec = this->get_executor();
+    auto trans_cpy = Fbcsr::create(exec, gko::transpose(this->get_size()),
+                                   this->get_num_stored_elements(), bs_);
+
+    exec->run(fbcsr::make_transpose(this, trans_cpy.get()));
+    return std::move(trans_cpy);
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Fbcsr<ValueType, IndexType>::conj_transpose() const
+{
+    auto exec = this->get_executor();
+    auto trans_cpy = Fbcsr::create(exec, gko::transpose(this->get_size()),
+                                   this->get_num_stored_elements(), bs_);
+
+    exec->run(fbcsr::make_conj_transpose(this, trans_cpy.get()));
+    return std::move(trans_cpy);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::sort_by_column_index() GKO_NOT_IMPLEMENTED;
```

Comment:
Is this too difficult to implement ? I have found this to be very useful/necessary in many cases.

Reply:
I was thinking of that too; I agree. But the current implementation for `Csr`, using the custom iterators, cannot simply be used for `Fbcsr`. It'll take some work to adapt the iterator interface properly. Or I could ignore that and just write a less elegant and slower implementation from scratch, without using `std::sort`.